File: android-starter/app/src/main/java/com/evplatform/driver/ui/screens/ChargingSessionScreen.kt
package com.evplatform.driver.ui.screens

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.evplatform.driver.model.ChargingSession
import com.evplatform.driver.model.InsuranceClaim
import com.evplatform.driver.model.SessionStatus
import com.evplatform.driver.network.ApiClient
import com.evplatform.driver.ui.theme.EVColor

/**
 * Live session screen. `session` is expected to update via SessionSocket
 * (WebSocket, fed by OCPP MeterValues) — not by polling — see ApiClient.kt.
 */
@Composable
fun ChargingSessionScreen(
    session: ChargingSession?,
    batteryCapacityKwh: Double,
    apiClient: ApiClient,
    onStop: () -> Unit
) {
    val percent = if (session != null && batteryCapacityKwh > 0)
        (session.energyKwh / batteryCapacityKwh).coerceIn(0.0, 1.0) else 0.0
    val isDone = session?.status in setOf(SessionStatus.COMPLETED, SessionStatus.FAILED, SessionStatus.STOPPED_REMOTELY)

    var claim by remember { mutableStateOf<InsuranceClaim?>(null) }

    // The instant a Guaranteed connector's session fails, insurance.py has already
    // auto-filed the claim server-side — this just fetches it to show the driver.
    LaunchedEffect(session?.status, session?.id) {
        if (session?.status == SessionStatus.FAILED) {
            claim = apiClient.claim(session.id)
        }
    }

    Column(
        Modifier.fillMaxSize().background(EVColor.bgDeep).padding(20.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Column(Modifier.fillMaxWidth()) {
            Text("LIVE SESSION", color = EVColor.textMuted, fontSize = 11.sp)
            Text("Charging", color = EVColor.textMain, fontSize = 26.sp, fontWeight = FontWeight.SemiBold)
        }

        Spacer(Modifier.height(24.dp))
        ChargeRing(percent = percent.toFloat(), status = session?.status ?: SessionStatus.PENDING)
        Spacer(Modifier.height(24.dp))

        Row(horizontalArrangement = Arrangement.spacedBy(10.dp), modifier = Modifier.fillMaxWidth()) {
            MiniStat("%.1f".format(session?.energyKwh ?: 0.0), "kWh delivered")
            MiniStat("₹${(session?.cost ?: 0.0).toInt()}", "cost so far")
            MiniStat("${(session?.powerKw ?: 0.0).toInt()} kW", "charge rate")
        }

        if (session?.status == SessionStatus.FAILED) {
            Spacer(Modifier.height(16.dp))
            FailBanner(claim)
        }

        if (!isDone) {
            Spacer(Modifier.height(20.dp))
            OutlinedButton(
                onClick = onStop,
                modifier = Modifier.fillMaxWidth().height(52.dp),
                shape = RoundedCornerShape(14.dp)
            ) { Text("Stop charging", color = EVColor.textMain) }
        }
    }
}

/**
 * Surfaces the guaranteed-charge insurance outcome — auto-filed server-side
 * the instant a Guaranteed connector's session fails (see
 * backend/app/services/insurance.py). No "file a claim" step for the driver.
 */
@Composable
private fun FailBanner(claim: InsuranceClaim?) {
    Column(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(14.dp))
            .background(EVColor.red.copy(alpha = 0.12f))
            .border(1.dp, EVColor.red, RoundedCornerShape(14.dp))
            .padding(14.dp)
    ) {
        Text("⚠ Station fault", color = EVColor.red, fontWeight = FontWeight.Bold, fontSize = 14.sp)
        Spacer(Modifier.height(4.dp))
        if (claim != null) {
            Text(
                "✓ ₹%.2f guaranteed-charge credit issued automatically".format(claim.creditAmount),
                color = EVColor.lime, fontSize = 13.sp
            )
        } else {
            Text("The connector faulted mid-session.", color = EVColor.textMuted, fontSize = 13.sp)
        }
    }
}

/** The signature element — a segmented gauge, not a generic progress bar. */
@Composable
private fun ChargeRing(percent: Float, status: SessionStatus) {
    val animated by animateFloatAsState(targetValue = percent, label = "chargeRing")
    val ringColor = when (status) {
        SessionStatus.FAILED -> EVColor.red
        SessionStatus.COMPLETED, SessionStatus.STOPPED_REMOTELY -> EVColor.lime
        else -> EVColor.amber
    }
    Box(Modifier.size(230.dp), contentAlignment = Alignment.Center) {
        Canvas(Modifier.size(230.dp)) {
            val stroke = Stroke(width = 14.dp.toPx(), cap = StrokeCap.Round)
            drawArc(EVColor.bgRaise, 0f, 360f, false, style = stroke, size = Size(size.width, size.height))
            drawArc(ringColor, -90f, 360f * animated.coerceAtLeast(0.02f), false, style = stroke, size = Size(size.width, size.height))
        }
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            Text("${(animated * 100).toInt()}%", color = EVColor.textMain, fontSize = 44.sp, fontWeight = FontWeight.Bold)
            Text("● ${status.name}", color = ringColor, fontSize = 11.sp)
        }
    }
}

@Composable
private fun RowScope.MiniStat(value: String, label: String) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = Modifier.weight(1f).clip(RoundedCornerShape(14.dp)).background(EVColor.bgPanel).padding(vertical = 12.dp)
    ) {
        Text(value, color = EVColor.textMain, fontWeight = FontWeight.Bold, fontSize = 18.sp)
        Text(label.uppercase(), color = EVColor.textMuted, fontSize = 9.sp)
    }
}

File: ios-starter/EVCharge/Screens/ChargingSessionScreen.swift
import SwiftUI

struct ChargingSessionScreen: View {
    let sessionId: UUID
    let batteryCapacityKwh: Double
    @StateObject private var socket = SessionSocket()
    @State private var claim: InsuranceClaim?

    private var percent: Double {
        guard let s = socket.session, batteryCapacityKwh > 0 else { return 0 }
        return min(s.energyKwh / batteryCapacityKwh, 1.0)
    }

    private var isDone: Bool {
        guard let status = socket.session?.status else { return false }
        return [.completed, .failed, .stoppedRemotely].contains(status)
    }

    var body: some View {
        VStack(spacing: 24) {
            VStack(alignment: .leading, spacing: 2) {
                Text("LIVE SESSION").font(EVTheme.Font.mono(11)).foregroundColor(EVTheme.textMuted)
                Text("Charging").font(EVTheme.Font.display(26))
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            ChargeRing(percent: percent, status: socket.session?.status ?? .pending)
                .frame(width: 230, height: 230)

            HStack(spacing: 10) {
                MiniStat(value: String(format: "%.1f", socket.session?.energyKwh ?? 0), label: "kWh delivered")
                MiniStat(value: "₹\(Int(socket.session?.cost ?? 0))", label: "cost so far")
                MiniStat(value: "\(Int(socket.session?.powerKw ?? 0)) kW", label: "charge rate")
            }

            if socket.session?.status == .failed {
                FailBanner(claim: claim)
            }

            if !isDone {
                Button(role: .destructive) {
                    Task { _ = try? await APIClient.shared.stopSession(id: sessionId) }
                } label: {
                    Text("Stop charging")
                        .frame(maxWidth: .infinity).padding()
                        .background(EVTheme.bgRaise).foregroundColor(EVTheme.textMain)
                        .clipShape(Capsule())
                        .overlay(Capsule().stroke(EVTheme.line))
                }
            }

            Spacer()
        }
        .padding(20)
        .background(EVTheme.bgDeep.ignoresSafeArea())
        .onAppear { socket.connect(sessionId: sessionId) }
        .onDisappear { socket.disconnect() }
        .onChange(of: socket.session?.status) { status in
            guard status == .failed else { return }
            Task { claim = try? await APIClient.shared.claim(sessionId: sessionId) }
        }
    }
}

/// Surfaces the guaranteed-charge insurance outcome — auto-filed server-side
/// the instant a Guaranteed connector's session fails (see
/// backend/app/services/insurance.py). No "file a claim" step for the driver.
private struct FailBanner: View {
    let claim: InsuranceClaim?

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text("⚠ Station fault").font(.system(size: 14, weight: .bold)).foregroundColor(EVTheme.red)
            if let claim {
                Text("✓ ₹\(String(format: "%.2f", claim.creditAmount)) guaranteed-charge credit issued automatically")
                    .font(EVTheme.Font.body(13)).foregroundColor(EVTheme.lime)
            } else {
                Text("The connector faulted mid-session.").font(EVTheme.Font.body(13)).foregroundColor(EVTheme.textMuted)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(14)
        .background(EVTheme.red.opacity(0.12))
        .overlay(RoundedRectangle(cornerRadius: EVTheme.Radius.panel).stroke(EVTheme.red))
        .clipShape(RoundedRectangle(cornerRadius: EVTheme.Radius.panel))
    }
}

/// The signature element: a segmented gauge, not a generic progress bar —
/// ticks read like an instrument, matching the web prototype's charge ring.
private struct ChargeRing: View {
    let percent: Double
    let status: SessionStatus

    private var ringColor: Color {
        switch status {
        case .failed: return EVTheme.red
        case .completed, .stoppedRemotely: return EVTheme.lime
        default: return EVTheme.amber
        }
    }

    var body: some View {
        ZStack {
            Circle().stroke(EVTheme.bgRaise, lineWidth: 14)
            Circle()
                .trim(from: 0, to: max(percent, 0.02))
                .stroke(ringColor, style: StrokeStyle(lineWidth: 14, lineCap: .round))
                .rotationEffect(.degrees(-90))
                .animation(.easeInOut(duration: 0.6), value: percent)

            VStack(spacing: 2) {
                HStack(alignment: .firstTextBaseline, spacing: 2) {
                    Text("\(Int(percent * 100))").font(EVTheme.Font.mono(44, weight: .bold))
                    Text("%").font(EVTheme.Font.mono(20)).foregroundColor(EVTheme.textMuted)
                }
                Text("● \(status.rawValue.uppercased())").font(EVTheme.Font.mono(11)).foregroundColor(ringColor)
            }
        }
    }
}

private struct MiniStat: View {
    let value: String; let label: String
    var body: some View {
        VStack(spacing: 4) {
            Text(value).font(EVTheme.Font.mono(18, weight: .bold)).foregroundColor(EVTheme.textMain)
            Text(label.uppercased()).font(.system(size: 9)).foregroundColor(EVTheme.textMuted)
        }
        .frame(maxWidth: .infinity).padding(.vertical, 12)
        .background(EVTheme.bgPanel).clipShape(RoundedRectangle(cornerRadius: EVTheme.Radius.element))
        .overlay(RoundedRectangle(cornerRadius: EVTheme.Radius.element).stroke(EVTheme.line))
    }
}
